### PR TITLE
r.sun: Fix parallel computation

### PR DESCRIPTION
**Pull request by aaronsms**
Currently with OpenMP support, the behavior is that the metadata displayed such as sunrise time, sunset time and solar altitude etc. is not updated upon the last parallelized computation iteration #1298. This PR fixes this by ensuring the `sunGeom`, `sunVarGeom`, `sunSlopeGeom` and `sunRadVar` are updated in the thread performing the last iteration. Furthemore, there are potential data races issues on global variables such as `linke_min` and `lat_max`. These are fixed with reduction clauses.

This is my first PR for this project (I'm still getting familiarized with the project), I would appreciate some feedback/confirmation on the expected behavior. I am going to include some test cases that fails the undesired behavior.

Thanks for reviewiing!

### CODE REVIEW

**Comment by petrasovaa** [comment]
Thanks for your contribution, I will try to look at this hopefully this week!

**Comment by aaronsms** [comment]
Hi after some reading of the manual, it seems to me the metadata fields are not directly exposed to the testsuites, for example fields like sunrise time, sunset time, daylight time. I might be wrong but my current understanding is that `assertRaster*` methods in TestCase module only checks for the primary key (incidence, beam_rad, glob_rad, insol_time etc.). Are they any methods that exposes the metadata fields currently or do you think there is a need to write additional methods to expose them, or perhaps the testing of metadata is not necessary?

Also, I am actually looking to join the GSOC, and that's actually how I bump into this issue. I have drafted a proposal and will finalize my proposal by this deadline. Thanks!

**Comment by aaronsms** [comment]
Hi after some reading of the manual, it seems to me the metadata fields are not directly exposed to the testsuites, for example fields like sunrise time, sunset time, daylight time. I might be wrong but my current understanding is that `assertRaster*` methods in TestCase module only checks for the primary key (incidence, beam_rad, glob_rad, insol_time etc.). Are they any methods that exposes the metadata fields currently or do you think there is a need to write additional methods to expose them, or perhaps the testing of metadata is not necessary?

Also, I am actually looking to join the GSOC, and that's actually how I bump into this issue. I have drafted a proposal and will finalize my proposal by this deadline. Thanks!

**Comment by ninsbl** [comment]
> Hi after some reading of the manual, it seems to me the metadata fields are not directly exposed to the testsuites, for example fields like sunrise time, sunset time, daylight time. I might be wrong but my current understanding is that `assertRaster*` methods in TestCase module only checks for the primary key (incidence, beam_rad, glob_rad, insol_time etc.). Are they any methods that exposes the metadata fields currently or do you think there is a need to write additional methods to expose them, or perhaps the testing of metadata is not necessary?

You could have a look at `assertModuleKeyValue` (https://grass.osgeo.org/grass-stable/manuals/libpython/gunittest.html?highlight=assert%20univar#gunittest.case.TestCase.assertModuleKeyValue) or 
`assertLooksLike` (https://grass.osgeo.org/grass-stable/manuals/libpython/gunittest.html?highlight=assert%20univar#gunittest.case.TestCase.assertModuleKeyValue)
You could create a proper reference (dict or text) containing the correct metadata with `r.info` and then use those in one of the test cases above...

> Also, I am actually looking to join the GSOC, and that's actually how I bump into this issue. I have drafted a proposal and will soonfinalize my proposal. Thanks!

Nice.

**Comment by aaronsms** [comment]
I have included a simple test case to check for metadata consistency on both the threaded and non-threaded case. For the previous test cases, I have also removed the restrictions on the threaded test cases.

**Comment by nilason** [comment]
I can confirm the result of the reproduction example at https://github.com/OSGeo/grass/issues/1298#issue-799956642
with this PR gives the expected result with OpenMP enabled:

```
 |     ----------------------------------------------------------------       |
 |     Day [1-365]:                              100                          |
 |     Local (solar) time (decimal hr.):         12.0000                      |
 |     Solar constant (W/m^2):                   1367.000000                  |
 |     Extraterrestrial irradiance (W/m^2):      1362.410121                  |
 |     Declination (rad):                        0.138928                     |
 |     Latitude min-max(deg):                    35.7328 - 35.7395            |
 |     Sunrise time (hr.):                       5.61                         |
 |     Sunset time (hr.):                        18.39                        |
 |     Daylight time (hr.):                      12.77                        |
 |     Solar altitude (deg):                     62.2205                      |
 |     Solar azimuth (deg):                      180.0000                     |
 |     Linke turbidity factor:                   3.0                          |
 |     Ground albedo:                            0.200                        |
 |     -----------------------------------------------------------------      |
```

Great work!

File: raster/r.sun/testsuite/test_rsun.py [hunk 1]
```diff
@@ -3,6 +3,97 @@
 from grass.gunittest.gmodules import SimpleModule
 
 
+class TestRSunMode3(TestCase):
```
Comment (petrasovaa) [inline raster/r.sun/testsuite/test_rsun.py:6]: This should be named something like TestRSunMode1Metadata. Mode 1 and 2 refer to computing irradiance vs irradiation, there is no mode 3.

**Review by petrasovaa — CHANGES_REQUESTED** [review]
This looks great, please just address the minor comment in the tests and I think this is ready to be merged.

File: raster/r.sun/testsuite/test_rsun.py (hunk 1 above)
Comment (aaronsms) [reply raster/r.sun/testsuite/test_rsun.py:6]: Okay got it!

**Review by petrasovaa — APPROVED** [review]

**Comment by neteler** [comment]
@petrasovaa cherry-picking into 7.8.7 fails due to merge conflicts...

**Pull request merged**